### PR TITLE
feat(wrapper): sending trace before function timeout is reached

### PR DESCRIPTION
**Pull request by liadgo**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,6 +57,7 @@
     "eslint-plugin-mocha": "^4.11.0",
     "eslint-plugin-react": "^7.11.0",
     "husky": "^1.1.0",
+    "lolex": "^3.0.0",
```
Comment (galbash) [inline package.json:60]: ?

File: package.json [hunk 2]
```diff
@@ -81,6 +82,7 @@
     "json.sortify": "^2.2.2",
     "md5": "^2.2.1",
     "node-sqlparser": "^1.0.4",
+    "nvm": "0.0.4",
```
Comment (galbash) [inline package.json]: this shouldn't be a dependency

File: src/event.js [hunk 3]
```diff
@@ -28,6 +28,14 @@ module.exports.setException = function setException(event, error) {
     }
 };
 
+/**
+ * Add timeout indication to a given event
+ * @param {proto.event_pb.Event} event The event the timeout is set on
+ */
+module.exports.setTimeout = function setTimeout(event) {
```
Comment (galbash) [inline src/event.js]: maybe markAsTimeout? (setTimeout is kinda reserved in node)

File: src/proto/error_code_pb.js [hunk 4]
```diff
@@ -18,7 +18,8 @@ goog.exportSymbol('proto.ErrorCode', null, global);
 proto.ErrorCode = {
   OK: 0,
   ERROR: 1,
-  EXCEPTION: 2
+  EXCEPTION: 2,
+  TIMEOUT: 3
```
Comment (galbash) [inline src/proto/error_code_pb.js:22]: did you change epsagon-proto as well?

File: src/wrappers/lambda.js [hunk 5]
```diff
@@ -106,6 +116,9 @@ function baseLambdaWrapper(
             // eslint-disable-next-line no-underscore-dangle
             process._events.beforeExit = originalBeforeExit;
 
+            // Mark trace as sent.
+            tracesSent = true;
```
Comment (galbash) [inline src/wrappers/lambda.js:117]: whats the difference between this and callbackCalled?

File: src/wrappers/lambda.js [hunk 6]
```diff
@@ -180,6 +193,11 @@ function baseLambdaWrapper(
         });
 
         try {
+            timeout = setTimeout(() => {
```
Comment (galbash) [inline src/wrappers/lambda.js]: timeoutHandler

File: src/wrappers/lambda.js (hunk 5 above)
Comment (galbash) [reply src/wrappers/lambda.js:117]: I think I understand the difference now, but it is confusing. Maybe unite `callbackCalled` and `traceSent` to `traceSent` or somea better name or something?

**Review by galbash — CHANGES_REQUESTED** [review]
take a look at the comments and tell me what you think

File: package.json (hunk 1 above)
Comment (liadgo) [reply package.json:60]: i'm using it in order to manipulate the test's clock.

File: src/event.js (hunk 3 above)
Comment (liadgo) [reply src/event.js]: K

File: src/proto/error_code_pb.js (hunk 4 above)
Comment (liadgo) [reply src/proto/error_code_pb.js:22]: :)

File: src/wrappers/lambda.js (hunk 6 above)
Comment (liadgo) [reply src/wrappers/lambda.js]: Kkk

File: src/wrappers/lambda.js [hunk 7]
```diff
@@ -83,6 +88,11 @@ function baseLambdaWrapper(
         };
 
         const handleUserExecutionDone = (error, result, sendSync) => {
+            clearTimeout(timeout);
+
+            if (tracesSent) {
```
Comment (galbash) [inline src/wrappers/lambda.js]: if with ||

**Review by galbash — APPROVED** [review]
just making sure you checked it a bit with sync/ async / callback/ context.fail/ context.succeed etc to make sure we didn't break anything right?

**Pull request merged**